### ISSUE TITLE
Track if plugin is doing recognizing work

### ISSUE THREAD

**Issue by 0ttik** [csdcorp/speech_to_text#224]
Ref: https://github.com/csdcorp/speech_to_text/issues/50
OS: Android 9

In situation like:
1. Start listening
2. Say something indistinguishable
3. Stop listening

I (not always, but pretty consistently) get no `event` with `event.eventType == SpeechRecognitionEventType.finalRecognitionEvent` or `event.eventType == SpeechRecognitionEventType.partialRecognitionEvent` firied in case of using `Stream<SpeechRecognitionEvent>`. So I have no good mechanism to detect if plugin is still recognizing or it failed to recognize beside waiting random amount of time (coz I imagine it can differ by connection quality) to conclude that plugin failed to recognize and to display such a message for user. 

More specifically I get only:
```
I/flutter (24844): event type: SpeechRecognitionEventType.statusChangeEvent
I/flutter (24844): event type: SpeechRecognitionEventType.statusChangeEvent
``` 
signaling that recording has started and then finished, but no other events. 

So solution is either make `event.eventType == SpeechRecognitionEventType.finalRecognitionEvent` obligatory after listening has finished or make a separate event for this or make a flag like `isRecognizing` to track this status. 

Now my workaround is to wait 3 seconds after listening has finished but it's kinda ugly because real recognition time differs.

**Comment by sowens-csd** [comment]
Thanks for posting. This is a very good point, there should be an event to help determine if listening is complete. I think in this case you want an event when the listener is no longer active, either because it has stopped or because it has timed out. Is that correct? 

In your step 3 above, do you call the `stop` method or the listener times out?

**Comment by 0ttik** [comment]
Thank you for fast reply. 
 
>I think in this case you want an event when the listener is no longer active, either because it has stopped or because it has timed out. Is that correct?

Yes. I imagine status events for `stop` method flow should be like: listening started -> listening finished -> recognition finished. I haven't looked into timing out events, but I assume something similar is needed there as well. 

>In your step 3 above, do you call the stop method or the listener times out?

I call `stop` method.

**Comment by sowens-csd** [comment]
Just working on this now. Looks like the `statusChangeEvent` is correct. It should have the `isListening` property false after completion. When `isListening` is false then it has finished listening. The partial and final events are only used when there are recognized words. 

Let me know if using the `statusChangeEvent` information resolves the issue or if you're looking for a different event.

**Comment by 0ttik** [comment]
Yes, I can determine if plugin is listening using `isListening` property.

But I can not detect if plugin is still _recognizing_ because after `isListening` is false I observe 2 cases:
1. Recognition finished and no partial or final events are fired, e.g. nothing recognized
2. Recognition finished and final event is fired: I can detect end of recognition

I want to detect when _recognition_ is finished in case №1.

**Comment by 0ttik** [comment]
@sowens-csd, does my previous post make any sense or there is existing/better way to determine when plugin is completed _recongnition_? Or maybe it's impossible at all?

For now I'm just setting arbitrary timeout like 2 seconds, but I'm not sure that it's the best solution, because I imagine that there could be a partial/final event even after this timeout (for example, in case of a slow internet connection), but I'm rejecting them because timeout has passed. 

Also I feel like sometimes I wait for too long when setting timeout for 2 seconds: plugin may already report that recognition has failed and I may stop bothering user with wait animation for whole 2 seconds.

**Comment by sowens-csd** [comment]
Sorry @0ttik I completely missed your earlier reply. Thanks for pinging me again. Yes, your previous post makes sense and I think I understand the issue better now. I'll have another look at the behaviour now.

**Comment by sowens-csd** [comment]
Just to make sure my understanding is correct, it sounds like you are getting the status change event before the final recognition event, at least in some cases? In other words you always get a status change with `isListening` false but then sometimes you get more speech events. The result is that you can't trust the status change event to indicate that speech recognition is complete. Is that correct?

**Comment by sowens-csd** [comment]
I just added a new event type the `doneEvent` which is fired once all recognition is complete. This set of changes should also ensure that only one `notListening` status is sent. There's also a change that should speed up recognition completion on iOS if you are playing sounds. If you have a chance and can try the version from the repo it would be helpful. Thanks!

**Comment by sowens-csd** [comment]
This change is available now in 5.0.0.
